### ISSUE TITLE
Microsoft Account OAuth2 sign-on fails when redirect URL is not under the website root

### ISSUE THREAD

**Issue by rxd2** [aspnet/AspNetKatana#203]
For example, https://my.site.com/signin-microsoft works but https://my.site.com/myapp/signin-microsoft does not.

GetOwinContext().Authentication.GetExternalLoginInfo() returns null. Callback URL contains error=access_denied parameter. Tracing shows Microsoft.Owin.Security.MicrosoftAccount.MicrosoftAccountAuthenticationMiddleware Error: 0 : Authentication failed messages.

This is broken in 4.0.0, starting with 3.1.0.

Reproducible both with localhost and the actual domain name.

**Comment by shahriarhusainy** [comment]
Scenario:

The application root is `https://localhost/OwinFbDemo`

Suppose `https://localhost/OwinFbDemo/securepage` is authenticated. `https://localhost/OwinDemo/Auth/LoginFB` points to the following action method under `AuthController.cs`:

```csharp
public ActionResult LoginFB()
{
	HttpContext.GetOwinContext().Authentication.Challenge(new 
		Microsoft.Owin.Security.AuthenticationProperties {
			RedirectUri = "/securepage"
		}, "Facebook");
	return new HttpUnauthorizedResult();
}
```

The issue is that after successful login through Facebook it will redirect to `https://localhost/securepage` instead of `https://localhost/OwinFbDemo/securepage`

**Comment by Tratcher** [comment]
You're able to work around it by setting RedirectUri = "/OwinFbDemo/securepage", correct?

The awkward bit is that if we fixed it anybody that had used this workaround would break.

**Comment by shahriarhusainy** [comment]
That worked. Thanks!

**Comment by rxd2** [comment]
The issue that shahriarhusainy described has obviously nothing to do with the issue I reported. Tratcher, did you have a chance to verify and/or take a look at what's going on?

Just to be clear, I don't manually set RedirectUrl. it's just that because my app is hosted not under the website root, but under virtual directory.

Thank you!

**Comment by Tratcher** [comment]
Virtual directory or sub application? They behave quite differently.

**Comment by rxd2** [comment]
Sub application, sorry

**Comment by Tratcher** [comment]
The error=access_denied parameter is fishy, that would be coming from the external MSA server. It sounds like an app registration problem in MSA.

**Comment by rxd2** [comment]
Not sure what you mean by "fishy". I just created a new application, under a different microsoft account. I have exactly the same problem. Works perfectly with https://localhost:44300/signin-microsoft. Fails with https://localhost:44300/booking/signin-microsoft. Is that a limitation of MSA? What can I do to help diagnose/fix this?

**Comment by muratg** [comment]
@rxd2 your issue seems to be a dup of https://github.com/aspnet/AspNetKatana/issues/212, not this one.

@shahriarhusainy glad your issue is resolved, thanks @Tratcher!

Closing this one.

**Comment by rxd2** [comment]
@muratg I don't agree, it's not a dup. #212 is maybe similar but not quite the same, My application does work with 4.0,0 as long as the redirect URL is a website root.

**Comment by muratg** [comment]
@rxd2 please file a separate issue as your issue doesn't seem to be related to this one either.

**Comment by rxd2** [comment]
@muratg Excuse me? It is my issue, I created it. How can it possibly be not "related to this issue"?

**Comment by muratg** [comment]
@rxd2 you're right, my apologies. @shahriarhusainy's issue was unrelated.

**Comment by rxd2** [comment]
@muratg No problem, thank you!

**Comment by Jimmyawesom** [comment]
Been trying to get a BOT and hacker out of my device for a year tried everything and that's out there. Nothing else there after my investments. It's gonna ruin me right now.
